design/hardware-classification-controller/support-for-new-parameters-hwcc-DiskAndNIC.md

```diff
@@ -0,0 +1,228 @@
+
+* Below is sample yaml for additional parameters in HardwareClassification.
+
+   ```yaml
+
+   apiVersion: metal3.io.sigs.k8s.io/v1alpha1
+   kind: HardwareClassification
+   metadata:
+     name: profile1
+   spec:
+     hardwareCharacteristics:
+       Disk:
+          minimumCount: 5
+          minimumIndividualSizeGB: 2200
+          diskSelector:
+            - hctl: "0:0:N:0"
+              rotational: false
+            - hctl: "0:0:N:0"
+              rotational: true
+            - hctl: "0:N:0:0"
+              rotational: true
+            - hctl: "N:0:0:0"
+              rotational: false
+       NIC:
+          minimumCount: 2
+          vendorSelector:
+            - vendor: 0x8086
+            - vendor: 0x15b3
+   ```
+
+* Update the existing Schema `HardwareClassification` by adding new
+  parameters "diskSelector" and "vendor" under Disk and NIC respectively.
+
+* Add validation for "diskSelector" and "vendor" parameters.
+
+* Under classification, from the `diskSelector` list we will filter
+  out the disk by using the `hctl` and `rotational` parameter. Then we
+  will compare the filtered disk for count and size parameter.
+  Classification of NVMe disk will be done using `rotational` and `model`
+  parameter. `model` parameter will contain NVMe keyword.
+  
+  For "vendor" parameter under `hardwareCharacteristics.nic.vendorSelector`,
+  we will use exact string match to classify host and if all vendors
+  present in host than host is classified.
+
+* Once the classification is completed, will add label if user has
+  provided any, otherwise will add default label as
+  `hardwareclassification.metal3.io/<profile-name>: matches`.
+
+### Risks and Mitigations
+
+This adds some complexity for user to define the rules in CR.
+Proper documentation will help to solve this issue.
+
+## Design Details
+
+### Work Items
+
+1. Add new parameters "diskSelector" under Disk and "vendor" under NIC
+   in file /api/v1alpha1/hardwareClassification_types.go.
+1. Update existing validation logic for new parameters introduced in
+   Disk and NIC.
+1. Extend existing comparison framework to support new parameters introduced
+   in Disk and NIC.
+1. Write unit tests for above implementation.
+
+### Dependencies
+
+Baremetal-Operator - To fetch BareMetal-host for classification.
+
+### Test Plan
+
+* Additional Unit tests for modified modules will be implemented.
+* Functional testing will be performed with respect to implemented
+  HardwareClassification CRD and controller.
+* Deployment & integration testing will be done.
+* We will carry out testing of HWCC on Dell hardware.
+
+## Alternatives
+
+In `diskSelector` field we are adding `hctl` and `rotational` fields
+only.
+
+Alternative of this is we can add all parameters of `rootDeviceHints`
+structure from baremetal-operator. It will be helpful to mirroring
+the structure for consistency.
```

Comment:
Could you add some of the justification that we discussed for why we made this choice?

Reply:
Added justification in proposal's Alternative section.
Kindly review.